beacon/light/range.go

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package light
+
+// Range represents a (possibly zero-length) range of integers (sync periods).
+type Range struct {
+	First, Next uint64
+}
+
+// IsEmpty returns true if the length of the range is zero.
+func (a Range) IsEmpty() bool {
+	return a.Next == a.First
+}
+
+// Includes returns true if the range includes the given period.
+func (a Range) Includes(period uint64) bool {
+	return period >= a.First && period < a.Next
+}
+
+// CanExpand returns true if the range includes or can be expanded with the given
+// period (either the range is empty or the given period is inside, right before or
+// right after the range).
+func (a Range) CanExpand(period uint64) bool {
```

Comment:
This exact check is needed in many places so I'd rather keep it, maybe change its name if there is a better suggestion. `canonicalStore` operates on items in continuous ranges and does not care any further about how it's used. It is sometimes expanded on both ends. The `CommitteeChain` uses three `canonicalStore`s so it also needs this check to check whether it can add a certain item.